GitExtensions/StaticDependencyInjection.cs

```diff
@@ -0,0 +1,118 @@
+﻿using GitCommands;
+using System;
+using System.IO;
+using System.Collections.Concurrent;
+
+namespace GitExtensions
+{
+
+    public class DIObjectInstance<T> where T : class, new()
+    {
+        private static object lockObj = new object();
+        private T _Inst;
+        public T Inst
+        {
+            get
+            {
+                if (_Inst == null)
+                {
+                    lock (lockObj)
+                    {
+                        if (_Inst == null)
+                        {
+                            _Inst = StaticDI.CreateInstance<T>(ClearInst);
+                        }
+                    }
+                }
+
+                return _Inst;
+            }
+        }
+
+        private void ClearInst()
+        {
+            lock (lockObj)
+            {
+                _Inst = null;
+            }
+        }
+    }
+
+    public interface IInstanceFactory
+    {
+        T CreateInstance<T>() where T : class, new();
+    }
+
+    public class NewInstanceFactory : IInstanceFactory
+    {
+        public T CreateInstance<T>() where T : class, new()
+        {
+            return new T();
+        }
+    }
+
+    public class StaticDI
```

Comment:
Thanks, it is always better to be more thorough.
It seems that I false assumed that the 3 dependencies are static/stateless. The all three services depends on IFileSystem. Changing this assumption, my concerns change as follow:
1. The dependencies should be passed down the tree through a multiparam constructor starting from the Main proc. 
2. We should create a separate service GitWorkingDirService to which FindGitWorkingDir and IsValidGitWorkingDir should be moved.
3. It should be handled at the Nsubstitute level.

 I would not introduce the ServiceLocator as it is in conflict with 1)

I will update this PR to comply with the updated assumption.